wallet: Fix wallet directory initialization

BrandonOdiwuor opened the pull request:
Fixes #16220, see [#16220 (comment)](https://github.com/bitcoin/bitcoin/issues/16220#issuecomment-1429962692)

- util/system datadir code should not care about wallets or look for or create any wallet paths
- If -walletdir is specified, wallet code should use that directory to locate and list wallets.
- If -walletdir is specified and path is not a directory, wallet init should return a fatal error and refuse to start
- If -walletdir is not specified, wallet code should use <datadir>/wallets as the directory to locate and list wallets.
- If -walletdir is not specified and <datadir>/wallets does not exist, wallet init code should create it
    - Before creating <datadir>/wallets wallet should call ListWalletDir ListDatabases on <datadir>
    - If list is empty wallet should create <datadir>/wallets as a directory, otherwise it should create it as a symlink pointing to .

BrandonOdiwuor committed:
Fix wallet directory initialization

util/system datadir code should not care about wallets or create any wallet paths
If -walletdir is specified, wallet init code should use that directory
If -walletdir is specified and path is not a directory, return a fatal error
If -walletdir is not specified, use <datadir>/wallets if it is directory.
If -walletdir is not specified and <datadir>/wallets does not exist:
<datadir>/wallets wallet should call ListWalletDir ListDatabases on <datadir>
If list is empty wallet should create <datadir>/wallets as a directory,
otherwise it should create it as a symlink pointing to .
discussion at https://github.com/bitcoin/bitcoin/issues/16220

BrandonOdiwuor commented:
@MarcoFalke not sure why this test is failing, since it passes locally
![Screenshot from 2023-09-26 13-07-20](https://github.com/bitcoin/bitcoin/assets/15610188/bc49f7a5-917e-4a4c-8423-601cc918ebc3)

fanquake commented:
cc @ryanofsky @achow101

ryanofsky commented:
This seems promising, but it isn't exactly implementing what's described in https://github.com/bitcoin/bitcoin/issues/16220#issuecomment-1429962692 or the PR description so it has a different set of advantages and disadvantages. 

The difference between code and PR description is that code is not implementing: _"If list is empty wallet should create `<datadir>/wallets` as a directory, otherwise it should create it as a symlink pointing to `.`"_. Because of this, the PR can't simplify the [`GetWalletDir` ](https://github.com/bitcoin/bitcoin/blob/04265ba9378efbd4c35b33390b1e5cf246d420a9/src/wallet/walletutil.cpp#L11-L31) function so it still needs to check existence `<datadir>/wallets` and can't just return it unconditionally.

Maybe this is ok, and the PR is still worthwhile, because it at least it simplifies the node code, even if it doesn't simplify the wallet code, and it makes the separation cleaner because node code no longer explicitly needs to create the "wallets" directory and can avoid creating it if the wallet is disabled.

Another consequence of not creating ` "wallets" -> "."` symlink is that startup might be a little slower in a data directory that has wallets in the top level, since `ListDatabases().empty()` will be checked every time on startup, instead of just one time when symlink doesn't exist.

Other notes about the PR:

- It is suspicious why wallet functional tests need to be changed. This change should be backwards compatible so it is surprising that tests should have to change in order to pass. It would be good to have an explanation for that.
- I'm not sure if we currently have a test explicitly written to cover loading a datadir with no `wallets/` subdirectory and wallets in the top-level directory. If there is not an existing test, probably one should be written to make sure code in this PR doesn't break and wallets can still be loaded there.
- This PR isn't removing qt code that creates the "wallets" directory: https://github.com/bitcoin/bitcoin/blob/bf32ae3969833c767b2c88d4c7f609a569c7effe/src/qt/intro.cpp#L245, which should no longer be necessary, after the other changes.
- I think it would be better if this PR did not add a `g_wallet_init_interface.InitWalletDir` function, and instead just changed existing `VerifyWallets` function which is already determining the wallet directory. Getting rid of it would be nicer just to have fewer init functions, but also this code doesn't belong in WalletInit class because with #10102 WalletInit wallet/init.cpp functions are run in the node process, not the wallet process.

BrandonOdiwuor commented:
Accidentally closed the PR

BrandonOdiwuor commented:
@ryanofsky 

I have updated the PR and moved the wallet directory initialization code to VerifyWallets()

I have also added tests for the various init scenarios and removed the remaining TryCreateDirectories from QT

The tests changed in this PR are a reflection of the changes in the logic of the wallet dir init code as suggested, such as the `wallet/` dir currently being created even if the data dir is not empty or the mainnet `wallet/` dir only being created when running mainnet

I removed the symlink code (see screenshot attached) since it makes tests in wallet_multiwallet.py fai
![Screenshot from 2023-10-18 16-46-18](https://github.com/bitcoin/bitcoin/assets/15610188/86483e01-d2e5-43e5-ae4b-5b70bb68efcc)
l

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -238,18 +238,6 @@ bool Intro::showIfNeeded(bool& did_show_intro, int64_t& prune_MiB)
                 /* Cancel clicked */
                 return false;
             }
-            dataDir = intro.getDataDirectory();
-            try {
-                if (TryCreateDirectories(GUIUtil::QStringToPath(dataDir))) {
-                    // If a new data directory has been created, make wallets subdirectory too
-                    TryCreateDirectories(GUIUtil::QStringToPath(dataDir) / "wallets");
-                }
-                break;
-            } catch (const fs::filesystem_error&) {
-                QMessageBox::critical(nullptr, PACKAGE_NAME,
-                    tr("Error: Specified data directory \"%1\" cannot be created.").arg(dataDir));
-                /* fall through, back to choosing screen */
-            }
```
In commit "Fix wallet directory initialization" (17b4955e4b1e4f02f00a948b432f18caa9b250a5)

Is it a mistake to delete this whole block of code? I'd expect only line 245 (`TryCreateDirectories("wallets")`) to be deleted

BrandonOdiwuor replied:
fixed

ryanofsky commented on test/functional/wallet_listtransactions.py:
```diff
@@ -234,8 +234,8 @@ def run_externally_generated_address_test(self):
         # refill keypool otherwise the second node wouldn't recognize addresses generated on the first nodes
         self.nodes[0].keypoolrefill(1000)
         self.stop_nodes()
-        wallet0 = os.path.join(self.nodes[0].chain_path, self.default_wallet_name, "wallet.dat")
```
In commit "Fix wallet directory initialization" (17b4955e4b1e4f02f00a948b432f18caa9b250a5)

Is there a reason this test is changing? I'd expect the change to be backwards compatible so existing tests should not need to change

BrandonOdiwuor replied:
@ryanofsky This is because of the change in wallet dir init logic. The `wallet/` dir is created when a new node is started and no wallet is detected in data_dir

ryanofsky commented on test/functional/wallet_reorgsrestore.py:
```diff
@@ -88,7 +88,7 @@ def run_test(self):
         # Node0 wallet file is loaded on longest sync'ed node1
         self.stop_node(1)
         self.nodes[0].backupwallet(self.nodes[0].datadir_path / 'wallet.bak')
-        shutil.copyfile(self.nodes[0].datadir_path / 'wallet.bak', self.nodes[1].chain_path / self.default_wallet_name / self.wallet_data_filename)
```
In commit "Fix wallet directory initialization" (17b4955e4b1e4f02f00a948b432f18caa9b250a5)

Is there a reason this test is changing? I'd expect the change to be backwards compatible so existing tests should not need to change

BrandonOdiwuor replied:
This is also because of the change in the wallet dir init logic. The `wallet/ `dir is created when a new node is started and no wallet is detected in data_dir

sedited reviewed (commented):
Concept ACK

achow101 commented:
The PR didn't seem to attract much attention in the past. Also, the issue seems not important enough right now to keep it sitting around idle in the list of open PRs.

Closing due to lack of interest.
